Add code demo from NAWI spring meeting

adam-a-a opened the pull request:
## Fixes/Resolves:

## Summary/Motivation:
Create a code demo via a Jupyter notebook for the NF DSPMDE model.
Also, bring in @avdudchenko's flowsheet example for NF with bypass and export to GUI.

**EDIT: removing flowsheet examples and ui code and only keeping notebook in this PR.

## Changes proposed in this PR:
- add jupyter notebook
- incorporate `total_hardness` into MCAS property model
- silence some warnings the popped up in the Jupyter notebook

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

adam-a-a committed:
add NF with bypass flowsheet with initial test; will need to fix errors subsequently

adam-a-a committed:
cleanup imports and unused code

adam-a-a committed:
remove more unused imports to pass pylint test

avdudchenko committed:
first update and tested flowsheet setup

added NF with and with out by pass, added tests for both flowsheets.

avdudchenko committed:
fixeed bypass test

fixed failing by pass test

avdudchenko committed:
include UI for nf and nf with by pass

avdudchenko committed:
Update nf.py

fixing lint issues.

avdudchenko committed:
more linting fixes

avdudchenko committed:
Update test_nf.py

Fixed nf test values.

adam-a-a commented on watertap/examples/flowsheets/nf_dspmde/nf.py:
```diff
@@ -0,0 +1,463 @@
+        units=pyunits.mg / pyunits.kg,
+        doc="Water Hardness as CO3",
+    )
+    stream.total_hardness.unfix()
```
Why do you need to unfix? After creating the Var, it should already be unfixed. Is there a special reason for this?

avdudchenko replied:
No real good reason. Its actually legacy as I copy paste same code to add a var and include the fix line...

MichaelPesce commented on watertap/examples/flowsheets/nf_dspmde/nf.py:
```diff
@@ -0,0 +1,473 @@
+    # add_objective(m)
+    print("Optimizing with {} DOFS".format(degrees_of_freedom(m)))
+    result = solver.solve(m, tee=True)
+    return result
```
```suggestion
def optimize(m, solver=None, **kwargs):
    if solver is None:
        solver = get_solver()
    # add_objective(m)
    print("Optimizing with {} DOFS".format(degrees_of_freedom(m)))
    result = solver.solve(m, tee=True)
    return result
```

MichaelPesce replied:
the parameter_sweep function requires this optimize function as an argument, but I'm not sure if I can send in the solver with it. so can we check for solver and initialize it inside this function in the case it's not provided?

adam-a-a committed:
Check if solver is none

Co-authored-by: MichaelPesce <35114629+MichaelPesce@users.noreply.github.com>

adam-a-a committed:
add get_solver in functions that use solve arg; cleanup

avdudchenko committed:
update to ui variables

MichaelPesce commented on watertap/examples/flowsheets/nf_dspmde/nf_with_bypass.py:
```diff
@@ -0,0 +1,190 @@
+    m.fs.by_pass_splitter.split_fraction[0, "bypass"].unfix()
+
+
+def optimize(m, solver):
```
```suggestion
def optimize(m, solver=None):
```

avdudchenko committed:
Update watertap/examples/flowsheets/nf_dspmde/nf_with_bypass.py

Co-authored-by: MichaelPesce <35114629+MichaelPesce@users.noreply.github.com>

adam-a-a committed:
add initial jup nb; step 7 is still old stuff from RO tutorial

avdudchenko committed:
fixed ui cost vars.

adam-a-a committed:
add png for j nb

adam-a-a committed:
some cleanup and rearrangement

adam-a-a committed:
hide scaling util warnings

adam-a-a committed:
Merge branch 'main' into springmeeting23

adam-a-a committed:
fix nf ui code to resolve errors

adam-a-a committed:
delete commented code

adam-a-a committed:
Merge branch 'main' into springmeeting23

adam-a-a committed:
Merge branch 'main' into springmeeting23

avdudchenko committed:
update nf hardnes stool

avdudchenko committed:
update to nf costing and assumptions

avdudchenko committed:
fixed tests

adam-a-a committed:
revise notebook; add total hardness to MCAS; cleanu[

adam-a-a committed:
Merge branch 'springmeeting23' of https://github.com/nawi-hub/proteuslib into springmeeting23

adam-a-a committed:
modify add_hardness function to only add max_hardness var and constraint

adam-a-a committed:
checkpoint for notebook; added feed, product blocks, pump; tried calculate_state to setup feed but longer than before

adam-a-a committed:
save after spring meeting presentation

adam-a-a committed:
fine-tune j nb

adam-a-a committed:
Merge branch 'main' into springmeeting23

adam-a-a commented on watertap/core/util/scaling.py:
```diff
@@ -24,20 +24,30 @@ def transform_property_constraints(self):
         var_str = p.name
         if p.method is not None and self.is_property_constructed(var_str):
             var = getattr(self, var_str)
+            msg = (
```
@MarcusHolly I modified your utility function to not show logger warnings for cases where a variable is fixed.

adam-a-a committed:
remove todo notes to test transform_property_constraints and add test to check that no logger warning is displayed, as expected

adam-a-a committed:
Merge branch 'springmeeting23' of https://github.com/nawi-hub/proteuslib into springmeeting23

adam-a-a committed:
fix nf flowsheet tests

adam-a-a committed:
add assert_optimal_termination since this was excluded and nf with bypass didn't solve

adam-a-a committed:
fix scaling for linux checks and get "optimal" solution

adam-a-a committed:
fix nf bypass ui linting errors

adam-a-a commented:
@avdudchenko @MichaelPesce @dangunter   - the NF with bypass flowsheet is failing the Linux tests which indicates a problem with scaling since Windows tests are passing. 

The main goal of this PR is to merge in the code demo (Jupyter notebook) from the Spring meeting. The NF and NF+bypass flowsheets are really just there for export to the GUI, but the GUI aspect of the demo will require other PRs that bring in everything that was shown in the demo, from what I understand. 

I propose to remove the NF with bypass flowsheet from this PR and make it a separate PR, where the issue can be resolved there. Any additional GUI-related PRs that seek to incorporate demo of the NF with bypass flowsheet would also depend on that PR. Any objections?

avdudchenko commented:
> 

Lets try to fix the scaling issue. I will take a pass at it tomorrow. if I cant' get it to work we can go from there.

adam-a-a commented:
> > 
> 
> Lets try to fix the scaling issue. I will take a pass at it tomorrow. if I cant' get it to work we can go from there.

OK- the only obstacle for me was that I have Windows and had trouble trying to resolve on Linux. I do have WSL and tried resolving there, where I did manage to get the flowsheet test to pass locally, but for some odd reason, the checks still fail here. I think my version of Linux may be old, so updating now and going to retry. If you can run on Linux, then would be happy for you to take a shot at it tomorrow.

bknueven committed:
running SequentialDecomposition for initialization; initializing costing (#1055)

Commit:
solves on intel macOS

adam-a-a committed:
remove nf bypass files and move to separate PR

adam-a-a commented:
@bknueven I removed what I considered to be the problematic flowsheet and associated files and moved to #1062 . However, not surprisingly, the macos check is failing for the regular old nf flowsheet. 

I can do a couple things.
1) move the nf flowsheet along to that separate PR (#1062)
2) add the requires_idaes_solver mark
3) give you a chance to look at this before I proceed with either (1) or (2)

I guess the 4th option could be for me to try to resolve the issue, but I'd rather focus time on more pressing matters.

bknueven commented:
> @bknueven I removed what I considered to be the problematic flowsheet and associated files and moved to #1062 . However, not surprisingly, the macos check is failing for the regular old nf flowsheet. 
> 
> 
> 
> I can do a couple things.
> 
> 1) move the nf flowsheet along to that separate PR (#1062)
> 
> 2) add the requires_idaes_solver mark
> 
> 3) give you a chance to look at this before I proceed with either (1) or (2)
> 
> 
> 
> I guess the 4th option could be for me to try to resolve the issue, but I'd rather focus time on more pressing matters.

Let's move forward with (1) or (2). I don't have an opinion on which is preferable.

adam-a-a committed:
cleanup commented code

adam-a-a committed:
run black

adam-a-a committed:
skip macos test

adam-a-a commented:
@bknueven skipping the macos test then.

@dangunter @MichaelPesce not sure how heavily this would affect GUI users subsequently.

adam-a-a committed:
run black

adam-a-a committed:
remove nf flowsheet and ui code

adam-a-a committed:
cleanup

adam-a-a committed:
Merge branch 'main' into springmeeting23

adam-a-a committed:
Merge branch 'main' into springmeeting23

adam-a-a committed:
Merge branch 'main' into springmeeting23

bknueven commented on watertap/property_models/multicomp_aq_sol_prop_pack.py:
```diff
@@ -2132,6 +2164,10 @@ def calculate_scaling_factors(self):
 
+        if self.is_property_constructed("total_hardness"):
+            if iscale.get_scaling_factor(self.total_hardness) is None:
+                sf = 1 / value(self.total_hardness)
```
Not a huge fan of this...

bknueven commented on watertap/unit_models/nanofiltration_DSPMDE_0D.py:
```diff
@@ -1498,12 +1495,11 @@ def initialize_build(
                 self._automate_rescale_variables()
         # ---------------------------------------------------------------------
         # Solve unit attempt 1
-        # self.eq_solute_solvent_flux.deactivate()
```
As an aside, it's interesting that this approach doesn't seem to work for this model. We use something similar in OARO, which was inspired by the IDAES heat exchanger initialization.

adam-a-a replied:
It may be helpful for the model; I don't remember rechecking after the fact. This was something that I had in place when I was still debugging the model at a more fundamental level. Hard to recall, but once I found the real issues with the model, I think I found that this was no longer needed, commented it out, but never deleted. We could try it out again to see if it helps with stability, but would prefer to reserve for another PR.

bknueven reviewed (approved):
I left some comments, but I don't think they should be addressed in this PR.

hunterbarber commented on watertap/property_models/multicomp_aq_sol_prop_pack.py:
```diff
@@ -1684,6 +1690,33 @@ def rule_elec_cond_phase(b, p):
+                sum(
+                    b.flow_mol_phase_comp["Liq", j]
+                    / b.flow_vol_phase["Liq"]
+                    * 100.0869
```
What is the 100.0869 parameter here? converting to CaCO3 equivalent?

Also curious of the decision to have the in-model units for hardness as milligrams, non SI and inconsistent with other mass use within the model. Even if it's just for reporting, wouldn't it be better to convert directly in the `.report()` function, in case this value is ever needed for other model util.

hunterbarber reviewed (approved):
Only had one comment, but also only viewed the raw text of the Jupyter Notebook, so if there's a way to view it in the proper display I can take another look in case there's something I missed.
